library/kotlin/src/io/envoyproxy/envoymobile/RequestBuilder.kt

```diff
@@ -0,0 +1,159 @@
+package io.envoyproxy.envoymobile
+
+import java.net.URL
+
+
+/**
+ * Builder used for constructing instances of `Request` types.
+ *
+ * @param url URL for the request.
+ * @param method Method for the request.
+ */
+class RequestBuilder(
+    val url: URL,
+    val method: RequestMethod
+) {
+  // Headers to send with the request.
+  // Multiple values for a given name are valid, and will be sent as comma-separated values.
+  private val headers: MutableMap<String, MutableList<String>> = mutableMapOf()
+
+  // Trailers to send with the request.
+  // Multiple values for a given name are valid, and will be sent as comma-separated values.
+  private val trailers: MutableMap<String, MutableList<String>> = mutableMapOf()
+
+  // Serialized data to send as the body of the request.
+  private var body: ByteArray? = null
+
+  // Retry policy to use for this request.
+  private var retryPolicy: RetryPolicy? = null
+
+  /**
+   * Serialized data to send as the body of the request.
+   */
+  fun addBody(body: ByteArray?): RequestBuilder {
+    this.body = body
+    return this
+  }
+
+  /**
+   * Add a retry policy to use for this request.
+   *
+   * @param retryPolicy the {@link io.envoyproxy.envoymobile.RetryPolicy} for this request.
+   * @return this builder.
+   */
+  fun addRetryPolicy(retryPolicy: RetryPolicy?): RequestBuilder {
+    this.retryPolicy = retryPolicy
+    return this
+  }
+
+  /**
+   * Append a value to the header key.
+   *
+   * @param name the header key.
+   * @param value the value associated to the header key.
+   * @return this builder.
+   */
+  fun addHeader(name: String, value: String): RequestBuilder {
+    if (headers.containsKey(name)) {
+      headers[name]!!.add(value)
+    }
+
+    headers[name] = mutableListOf(value)
```

Comment:
This is a bug - I think you need to early-return above to avoid re-setting the array and clearing previously present values